collector/xfs_linux.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+	"github.com/prometheus/procfs/xfs"
+)
+
+// An xfsCollector is a Collector which gathers metrics from XFS filesystems.
+type xfsCollector struct {
+	fs sysfs.FS
+}
+
+func init() {
+	Factories["xfs"] = NewXFSCollector
+}
+
+// NewXFSCollector returns a new Collector exposing XFS statistics.
+func NewXFSCollector() (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %v", err)
+	}
+
+	return &xfsCollector{
+		fs: fs,
+	}, nil
+}
+
+// Update implements Collector.
+func (c *xfsCollector) Update(ch chan<- prometheus.Metric) error {
+	stats, err := c.fs.XFSStats()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve XFS stats: %v", err)
+	}
+
+	for _, s := range stats {
+		c.updateXFSStats(ch, s)
+	}
+
+	return nil
+}
+
+// updateXFSStats collects statistics for a single XFS filesystem.
+func (c *xfsCollector) updateXFSStats(ch chan<- prometheus.Metric, s *xfs.Stats) {
+	const (
+		subsystem = "xfs"
+	)
+
+	var (
+		labels = []string{"device"}
+	)
+
+	metrics := []struct {
+		name  string
+		desc  string
+		field float64
+	}{
+		{
+			name:  "extent_allocation_extents_allocated_total",
```

Comment:
Our usual naming pattern for counters is `namespace_thing_action_total` where action is a noun, like `extent_allocations_total`, `btree_comparisons_total`, etc. Your current one is consistent as well, I'll leave it to @SuperQ.